### DISCUSSION TITLE
Util Import functions not creating id for created Rules / Groups

### DISCUSSION THREAD

**Discussion by emmalcg** [react-querybuilder/react-querybuilder#542] [Q&A] [answered]
Hi! I am wondering if there is a recommended approach for adding `id` property to rules / groups that are created through import utils like `parseJsonLogic()`. I am facing react key errors as the Rule / Group `id` is being used by reactQueryBuilder it seems to create unique `key` props for these components. Without the unique id being added from `parseJsonLogic()`, I am running into react key errors in the console.
Are there any plans to add the id within these until import functions?

I also tried a custom function after the parseJsonLogic and before setting the query state, but am running into a ton of type errors

  const addIdsToRules = useCallback((node: DefaultRuleGroupType) => {
    if (node.id === undefined) {
      node.id = generateIdString();
    }

    if (node.rules && node.rules.length) {
      node.rules.forEach(addIdsToRules);
    }

    if (node.combinator === 'or' || node.combinator === 'and') {
      node.rules.forEach(addIdsToRules);
    }
    return node;
  }, []);

**Comment by jakeboone02** [comment, marked as answer]
No, I haven't considered adding `id`s to queries the parsers generate. In the meantime you could use `prepareRuleGroup` (undocumented utility function) to apply an `id` to your query and all its rules and subgroups. Maybe something like this:

```ts
const query = parseSQL("firstName = 'Steve'");
const queryWithIDs = prepareRuleGroup(query);
// ...or using a custom idGenerator:
// const queryWithIDs = prepareRuleGroup(query, { idGenerator });

/*
`queryWithIDs` would be:
{
  id: 'SOME-UUID',
  combinator: 'and',
  rules: [{
    id: 'DIFFERENT-UUID',
    field: 'firstName',
    operator: '=',
    value: 'Steve',
  }],
}
*/
```

**Reply by emmalcg** [reply]
Thanks so much! prepareRuleGroup works perfectly 👏

**Reply by emmalcg** [reply]
Hi @jakeboone02 just wanted to flag that prepareRuleGroup always adds a custom id and never a `root` id. Not sure if this is expected or not from this function but wanted to flag just in case :) 
Created a work around where i add the id of `root` after import and before calling `prepareRuleGroup`

**Reply by jakeboone02** [reply]
Do you mean it doesn't add an `id` to the group you pass in, only the sub-rules/groups? It should definitely be doing that if it isn't. Can you provide some example code that demonstrates the issue?

**Reply by emmalcg** [reply]
@jakeboone02 it does add an id to every group - i just noticed it is not the id of "root" and instead a randomized id. However when you make a query from scratch (without import) the id is "root". 
I was using the id of "root" to make some visual changes to the "root group", but realized the id is not always root because of the import, so did something else. Not a bug or immediate issue - just wanted to flag because of the discrepancies in the data structure but maybe that is intentional!

**Reply by jakeboone02** [reply]
I'm not sure what you mean by "make a query from scratch." `"root"` isn't hard-coded anywhere in the RQB source (except for some tests that don't make it into the build).

If you're looking to style the root group, some CSS like this should work:

```css
.queryBuilder > .ruleGroup {
  /* styles here only affect root group(s) */
}
```

**Reply by emmalcg** [reply]
cool thanks! yeah i just noticed the id of the query object was always 'root' except for on import - but this isn't a problem or affecting anything negatively!
